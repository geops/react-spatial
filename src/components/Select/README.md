#

This demonstrates the use of Select.

```jsx
import React from  'react'
import Select from 'react-spatial/components/Select';

class SelectExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = { value: 'ny' };
  }

  render() {
    return (
      <div className="tm-select-example">
        <Select
          options={[
            { label: 'New york', value: 'ny' },
            { label: 'Los angeles', value: 'ls' },
            { label: 'Orlando', value: 'orl' },
          ]}
          value={this.state.value}
          onChange={(e, value) => {
            this.setState({ value: value });
          }}
        />
        <div>You selected "{this.state.value}".</div>
      </div>
    );
  }
}

<SelectExample />;
```
