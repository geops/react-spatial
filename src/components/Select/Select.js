import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  /**
   * CSS class of the select.
   */
  className: PropTypes.string,

  /**
   * Array of values to use in the select element.
   */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      label: PropTypes.string,
    }),
  ).isRequired,

  /**
   * Selected option object.
   */
  value: PropTypes.string,

  /**
   * Function triggered on select's change event.
   * @param {Event} event The change event object.
   */
  onChange: PropTypes.func.isRequired,
};

const defaultProps = {
  value: undefined,
  className: 'tm-select',
};

/**
 * This component displays a simple select HTML element.
 */
const Select = props => {
  const { value, className, options, onChange } = props;
  return (
    <select
      className={className}
      value={value}
      onChange={evt => onChange(evt, evt.target.value)}
    >
      {(options || []).map(o => (
        <option key={o.value} value={o.value}>
          {o.label}
        </option>
      ))}
    </select>
  );
};

Select.propTypes = propTypes;
Select.defaultProps = defaultProps;

export default Select;
