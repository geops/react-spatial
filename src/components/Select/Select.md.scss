.tm-select-example {
  width: 300px;

  select {
    padding: 10px;
  }

  div {
    padding: 10px 0;
  }
}
