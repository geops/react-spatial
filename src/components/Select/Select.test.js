import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Select from './Select';

configure({ adapter: new Adapter() });

describe('Select', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('log error if required properties are not set.', () => {
      shallow(<Select />);
      expect(spy).toHaveBeenCalledTimes(2);
    });

    test('matches snapshot', () => {
      const component = renderer.create(<Select />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when properties are set', () => {
    const options = [
      {
        value: 'foo',
        label: 'qux',
      },
      {
        value: 'bar',
        label: 'quux',
      },
      {
        value: 'baz',
        label: 'corge',
      },
    ];

    const onChange = () => {};

    test('matches snapshot', () => {
      const component = renderer.create(
        <Select options={options} value="bar" onChange={() => {}} />,
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('executes onChange property function', () => {
      const onChangeMock = jest.fn(onChange);
      const wrapper = shallow(
        <Select options={options} value="baz" onChange={onChangeMock} />,
      );
      expect(onChangeMock).toHaveBeenCalledTimes(0);
      wrapper.find('select').simulate('change', { target: { value: 'foo' } });
      expect(onChangeMock).toHaveBeenCalledTimes(1);
    });
  });
});
