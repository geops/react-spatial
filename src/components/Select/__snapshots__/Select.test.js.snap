// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select when no properties are set matches snapshot 1`] = `
<select
  className="tm-select"
  onChange={[Function]}
/>
`;

exports[`Select when properties are set matches snapshot 1`] = `
<select
  className="tm-select"
  onChange={[Function]}
  value="bar"
>
  <option
    value="foo"
  >
    qux
  </option>
  <option
    value="bar"
  >
    quux
  </option>
  <option
    value="baz"
  >
    corge
  </option>
</select>
`;
